### PR TITLE
"logarthmic" to "logarithmic", improve readability, test 2D patterns and optimal IB

### PR DESCRIPTION
**Pull request by chriscoey**
- change "logarthmic" to "logarithmic"
- slightly improve readability/consistency, especially of tests
- test 2D patterns
- test optimal IB scheme (tragically slow)

### CODE REVIEW

**Comment by chriscoey** [comment]
@joehuchette optimal IB scheme is extremely slow to calculate (even for 4 breakpoints).. is it ever worthwhile?

also, where do the values for `z` in your tests come from? they don't match the true function values at the points, though they are the correct optimal solutions for the models constructed. are you just checking for consistency across the different formulations?

**Comment by chriscoey** [comment]
if optimal IB scheme test takes too long on travis, I will remove it. quite absurd how slow it is.

edit: reduced it to 3x3 breakpoints - works ok

File: src/jump.jl [hunk 1]
```diff
@@ -120,10 +124,10 @@ function sos2_cc_formulation!(m::JuMP.Model, λ)
         JuMP.@constraint(m, λ[i] ≤ y[i-1] + y[i])
     end
     JuMP.@constraint(m, λ[n] ≤ y[n-1])
-    nothing
+    return nothing
 end
 
-function sos2_mc_formulation!(m::JuMP.Model, λ)
+function sos2_mc_formulation!(m::JuMP.Model, λ) # not currently used
```
Comment (chriscoey) [inline src/jump.jl]: @joehuchette this is here but not used. can it be commented out?

**Comment by chriscoey** [comment]
I'm satisfied at this point

File: src/jump.jl [hunk 2]
```diff
@@ -611,332 +608,339 @@ function piecewiselinear(m::JuMP.Model, x₁::VarOrAff, x₂::VarOrAff, pwl::Biv
...
+            elseif pattern == :K1
+                # assumption is that the triangulation is uniform, as defined in types.jl
+                w = JuMP.@variable(m, [1:2], Bin, basename="w_$counter")
+                JuMP.@constraints(m, begin
+                    sum(λ[tx,ty] for tx in 1:nˣ, ty in 1:nʸ if mod(tx,2) == mod(ty,2) && (tx+ty) in 2:4:(nˣ+nʸ)) ≤ w[1]
+                    sum(λ[tx,ty] for tx in 1:nˣ, ty in 1:nʸ if mod(tx,2) == mod(ty,2) && (tx+ty) in 4:4:(nˣ+nʸ)) ≤ 1 - w[1]
+                    sum(λ[tx,ty] for tx in 1:nˣ, ty in 1:nʸ if mod(tx,2) != mod(ty,2) && (tx+ty) in 3:4:(nˣ+nʸ)) ≤ w[2]
+                    sum(λ[tx,ty] for tx in 1:nˣ, ty in 1:nʸ if mod(tx,2) != mod(ty,2) && (tx+ty) in 5:4:(nˣ+nʸ)) ≤ 1 - w[2]
+                end)
+            # elseif pattern == :OptimalTriangleSelection
```
Comment (joehuchette) [inline src/jump.jl]: Why is this commented out?

**Review by joehuchette — COMMENTED** [review]
Looks fine to me, but the diff for the triangulation formulations is hard to read. Did you change that code?

**Comment by chriscoey** [comment]
I commented out 3 parts of code in src/jump.jl:
- sos2 MC formulation
- optimal triangle selection pattern
- stencil pattern

because they all appeared to me to be not useable/accessible. imo it's better (and less confusing) to have untestable dead code commented out.

the two patterns above are not recognized in BivariatePWLFunction so it fails without an error message on https://github.com/joehuchette/PiecewiseLinearOpt.jl/blob/184ea708cd3d3869a95cd62b9dcb1923d03a1a77/src/types.jl#L101

**Comment by chriscoey** [comment]
I only made a few small aesthetic changes, mostly to comments, in the triangulation formulations

**Comment by joehuchette** [comment]
They're accessible if you change the ``pattern`` field after construction; I had research code that used this. The interface to specify the formulation needs to be rewritten (on my TODO list), but the code is not dead and shouldn't be removed.

**Comment by chriscoey** [comment]
OK so it sounds like a simple fix to allow the two extra patterns, right? a few extra lines in BivariatePWLFunction. we can do it here and then cover the two patterns with unit tests. 

which pattern did you need to specify before you could later change the pattern to stencil? and same question for triangle selection.

**Comment by joehuchette** [comment]
Yes, it should be a simple change to support them "officially".

I would generate the triangulation randomly, then change the ``pattern`` field manually. The conflation here between structure (random) and formulation (e.g. optimal triangle selection) is not great, and one of the reasons I would like to change the interface.

**Comment by chriscoey** [comment]
OK I uncommented. can you open an issue for the expected interface change

**Comment by joehuchette** [comment]
Thanks! #28 is for the interface change, take a look and let me know your thoughts.

**Pull request merged**